File: src/main/java/com/letgo/book/domain/BookId.java
package com.letgo.book.domain;

import java.util.UUID;

final public class BookId {
    private final UUID value;

    private BookId(UUID value) {
        this.value = value;
    }

    public static BookId create() {
        return new BookId(UUID.randomUUID());
    }

    public static BookId create(String value) {
        return new BookId(UUID.fromString(value));
    }

    public String value() {
        return value.toString();
    }

    @Override
    public boolean equals(Object anObject) {
        return anObject instanceof BookId && value().equals(((BookId) anObject).value());
    }

    @Override
    public int hashCode() {
        return value().hashCode();
    }
}

File: src/main/java/com/letgo/book/domain/BookId.kt
package com.letgo.book.domain

import java.io.Serializable
import java.util.*

data class BookId private constructor(
    private val value: UUID
) : Serializable {
    fun value(): String {
        return value.toString()
    }

    companion object {
        fun create(value: String): BookId {
            return BookId(UUID.fromString(value))
        }
    }
}
